File: demo/p3/node/p3_server.js
const readline = require('readline');
const path = require('path');
const { FRPCClient, TunnelType } = require('../../../bindings/nodejs/frpc_node');

if (process.argv.length < 4) {
    console.log(`Usage: node ${path.basename(__filename)} <server_addr> <server_port>`);
    process.exit(1);
}

const host = process.argv[2];
const port = parseInt(process.argv[3]);

console.log(`Starting Node.js STCP Server connecting to ${host}:${port}...`);

const client = new FRPCClient(host, port, "test_token", { useEncryption: true });

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    prompt: '> '
});

const tunnel = client.createTunnel(TunnelType.STCP_SERVER, "p3_test_stcp", {
    secretKey: "p3_test_secret",
    localAddr: "127.0.0.1",
    localPort: 0,
    onData: (data) => {
        const msg = data.toString('utf8').trim();
        // Clear current line to avoid mess with prompt
        if (process.stdout.clearLine && process.stdout.cursorTo) {
            process.stdout.clearLine(0);
            process.stdout.cursorTo(0);
        }
        console.log(`[Received] ${msg}`);
        rl.prompt(true);
    },
    onConnection: (connected, error) => {
        if (process.stdout.clearLine && process.stdout.cursorTo) {
            process.stdout.clearLine(0);
            process.stdout.cursorTo(0);
        }
        const status = connected ? "Connected" : "Disconnected";
        console.log(`[Status] ${status} (error: ${error})`);
        rl.prompt(true);
    }
});

client.connect();
tunnel.start();

console.log("Server started. Waiting for visitors...");
rl.prompt();

rl.on('line', (line) => {
    line = line.trim();
    if (line === 'quit') {
        client.disconnect();
        process.exit(0);
    }
    if (line) {
        try {
            tunnel.sendData(`[Server] ${line}`);
        } catch (e) {
            console.error(`Send failed: ${e.message}`);
        }
    }
    rl.prompt();
});

File: demo/p3/rust/src/bin/p3_server.rs
use frpc::{FrpcClient, TunnelConfig, TunnelType, TunnelEventHandler};
use std::sync::Arc;
use std::io::{self, Write};
use std::env;

struct ServerHandler;

impl TunnelEventHandler for ServerHandler {
    fn on_data(&self, data: &[u8]) {
        let msg = String::from_utf8_lossy(data);
        print!("\r\x1b[K"); // Clear line
        println!("[Received] {}", msg.trim());
        print!("> ");
        io::stdout().flush().unwrap();
    }

    fn on_connection(&self, connected: bool, error_code: i32) {
        print!("\r\x1b[K");
        let status = if connected { "Connected" } else { "Disconnected" };
        println!("[Status] {} (error: {})", status, error_code);
        print!("> ");
        io::stdout().flush().unwrap();
    }
}

fn main() -> anyhow::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 3 {
        eprintln!("Usage: {} <server_addr> <server_port>", args[0]);
        std::process::exit(1);
    }
    let host = &args[1];
    let port: u16 = args[2].parse()?;

    println!("Starting Rust STCP Server connecting to {}:{}...", host, port);

    let mut client = FrpcClient::new(host, port, Some("test_token"))?;
    client.set_encryption(true);

    let handler = Arc::new(ServerHandler);

    let config = TunnelConfig {
        tunnel_type: TunnelType::StcpServer,
        tunnel_name: "p3_test_stcp".to_string(),
        secret_key: Some("p3_test_secret".to_string()),
        local_addr: Some("127.0.0.1".to_string()),
        local_port: Some(0),
        remote_name: None,
        bind_addr: None,
        bind_port: None,
    };

    let tunnel = client.create_tunnel(config, Some(handler))?;
    client.connect()?;
    tunnel.start()?;

    println!("Server started. Waiting for visitors...");
    println!("Type message and press Enter to send (or 'quit' to exit):");
    print!("> ");
    io::stdout().flush()?;

    let stdin = io::stdin();
    let mut line = String::new();
    while stdin.read_line(&mut line)? > 0 {
        let trimmed = line.trim();
        if trimmed == "quit" {
            break;
        }
        if !trimmed.is_empty() {
            let msg = format!("[Server] {}", trimmed);
            tunnel.send_data(msg.as_bytes())?;
        }
        print!("> ");
        io::stdout().flush()?;
        line.clear();
    }

    client.disconnect()?;
    Ok(())
}
